File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/entity/component/StatusComponentActor.java
package net.bestia.zoneserver.actor.entity.component;

import java.util.Objects;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import net.bestia.entity.component.StatusService;
import net.bestia.model.domain.ConditionValues;
import net.bestia.zoneserver.actor.BestiaPeriodicTerminatingActor;

/**
 * The actor checks an entity with a status component attached and will
 * periodically calculate (usually every second) the hp and mana regeneration.
 * It will then update the current Mana and current HP values.
 * 
 * @author Thomas Felix
 *
 */
@Component
@Scope("prototype")
public class StatusComponentActor extends BestiaPeriodicTerminatingActor {

	public static final String NAME = "statusComponent";
	
	private final long entityId;
	private final StatusService statusService;

	private float manaIncrement;
	private float healthIncrement;

	@Autowired
	public StatusComponentActor(StatusService statusService, long entityId) {

		this.statusService = Objects.requireNonNull(statusService);
		this.entityId = entityId;

		startInterval(StatusService.REGENERATION_TICK_RATE_MS);
	}

	@Override
	protected void onTick() {

		try {

			healthIncrement += statusService.getHealthTick(entityId);
			manaIncrement += statusService.getManaTick(entityId);

			ConditionValues sval = statusService.getConditionalValues(entityId).orElseThrow(IllegalArgumentException::new);

			if (healthIncrement > 1) {
				
				// Update health status.
				final int hpRound = (int) healthIncrement;
				healthIncrement -= hpRound;
				sval.addHealth(hpRound);

			} 

			if (manaIncrement > 1) {
				
				// Update mana.
				final int manaRound = (int) manaIncrement;
				manaIncrement -= manaRound;
				sval.addMana(manaRound);

			}
			
			statusService.save(entityId, sval);

		} catch (IllegalArgumentException e) {
			// Could not tick regeneration for this entity id. Probably no
			// status component attached.
			// Terminating.
			context().stop(getSelf());
		}
	}

	@Override
	protected void handleMessage(Object message) {
		// no op.
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/entity/component/StatusComponentActor.kt
package net.bestia.zoneserver.actor.entity.component

import akka.japi.pf.ReceiveBuilder
import net.bestia.zoneserver.actor.ActorComponent
import net.bestia.zoneserver.entity.component.AllEntitiesStatusComponentFactory
import net.bestia.zoneserver.entity.component.EquipComponent
import net.bestia.zoneserver.entity.component.LevelComponent
import net.bestia.zoneserver.entity.component.StatusComponent

@ActorComponent(StatusComponent::class)
class StatusComponentActor(
    component: StatusComponent,
    val statusComponentFactory: AllEntitiesStatusComponentFactory
) : ComponentActor<StatusComponent>(component) {

  override fun createReceive(builder: ReceiveBuilder) {
    builder
        .match(LevelComponent::class.java) { calculateStatusValues() }
        .match(EquipComponent::class.java) { calculateStatusValues() }
  }

  override fun preStart() {
    createComponentUpdateSubscription(LevelComponent::class.java)
    createComponentUpdateSubscription(EquipComponent::class.java)
  }

  private fun calculateStatusValues() {
    requestOwnerEntity {
      component = statusComponentFactory.buildComponent(it)
    }
  }

  companion object {
    const val NAME = "statusComponent"
  }
}
